rayon-core/src/thread_pool/test.rs

```diff
@@ -57,3 +61,59 @@ fn sleeper_stop() {
     // should lead to worker threads stopping
     registry.wait_until_stopped();
 }
+
+fn count_handler() -> (Arc<AtomicUsize>, ::StartHandler) {
+    let count = Arc::new(AtomicUsize::new(0));
+    (count.clone(), Arc::new(move |_| { count.fetch_add(1, Ordering::SeqCst); }))
+}
+
+#[test]
+fn failed_thread_stack() {
+    use std::{thread, time};
+
+    let (start_count, start_handler) = count_handler();
+    let (exit_count, exit_handler) = count_handler();
+    let config = Configuration::new()
+        .set_stack_size(::std::usize::MAX)
+        .set_start_handler(start_handler)
+        .set_exit_handler(exit_handler);
+
+    let pool = ThreadPool::new(config);
+    assert!(pool.is_err(), "thread stack should have failed!");
+
+    // Give time for the internal cleanup to terminate.
+    thread::sleep(time::Duration::from_secs(1));
+
+    // With an impossible stack, we don't expect to see any threads.
+    assert_eq!(0, start_count.load(Ordering::SeqCst));
+    assert_eq!(0, exit_count.load(Ordering::SeqCst));
+}
+
+#[test]
+fn panic_thread_name() {
+    use std::{thread, time};
+
+    let (start_count, start_handler) = count_handler();
+    let (exit_count, exit_handler) = count_handler();
+    let config = Configuration::new()
+        .set_num_threads(10)
+        .set_start_handler(start_handler)
+        .set_exit_handler(exit_handler)
+        .set_thread_name(|i| {
+                             if i >= 5 {
+                                 panic!();
+                             }
+                             format!("panic_thread_name#{}", i)
+                         });
+
+    let pool = unwind::halt_unwinding(|| ThreadPool::new(config));
+    assert!(pool.is_err(), "thread-name panic should propagate!");
+
+    // Give time for the internal cleanup to terminate.
+    thread::sleep(time::Duration::from_secs(1));
```

Comment:
I think it'd be better to have the exit handlers send messages to a shared queue (i.e., a channel), so that we can just read from the channel and verify that we get the expected number of responses. This won't be sensitive to timing.